File: benchmarks/ts-bench/src/tests/throughput.ts
import type { Config } from '../config.js';
import type { TierResult } from '../report.js';
import * as protocol from '../protocol.js';
import { printHeader, fmtCount, fmtBytesPerSec } from '../stats.js';
import { generateBenchToken } from '../jwt.js';
import { tiersFor } from '../config.js';
import type WebSocket from 'ws';

export async function run(config: Config): Promise<TierResult[]> {
  printHeader(
    'TEST 3: Throughput Saturation',
    `All connections send as fast as possible for ${config.duration}s. Finds peak msg/s.`
  );

  const tiers = tiersFor(config, 'throughput');
  const token = generateBenchToken(Buffer.from(config.secret), 'bench-user');
  const payload = protocol.buildPayload(175);
  const msgBuf = Buffer.from(payload);
  const results: TierResult[] = [];

  for (const n of tiers) {
    console.log(`\n  --- ${n} connections ---`);

    const connections = await protocol.connectBatch(
      config.host, config.port, token, n, config.batchSize,
      'compression=false&protocol_version=1'
    );

    if (connections.length === 0) {
      console.log('    FAILED: no connections');
      continue;
    }

    const actual = connections.length;
    const durationMs = config.duration * 1000;

    const totalStart = process.hrtime.bigint();
    const deadline = Date.now() + durationMs;

    // Send as fast as possible on all connections
    const sendPromises = connections.map((ws) => sendLoop(ws, msgBuf, deadline));
    const sendResults = await Promise.allSettled(sendPromises);

    const totalSecs = Number(process.hrtime.bigint() - totalStart) / 1_000_000_000;

    let totalMessages = 0;
    let totalBytes = 0;
    const wsList: protocol.WsConnection[] = [];

    for (const r of sendResults) {
      if (r.status === 'fulfilled') {
        totalMessages += r.value.count;
        totalBytes += r.value.bytes;
        wsList.push(r.value.ws);
      }
    }

    const msgsPerSec = totalMessages / totalSecs;
    const bytesPerSec = totalBytes / totalSecs;
    const perConn = actual > 0 ? Math.round(msgsPerSec / actual) : 0;

    console.log(`    Duration: ${totalSecs.toFixed(2)}s`);
    console.log(
      `    Messages: ${totalMessages.toLocaleString()} (${fmtCount(Math.round(msgsPerSec))}/s)`
    );
    console.log(`    Bandwidth: ${fmtBytesPerSec(bytesPerSec)}`);
    console.log(`    Per-connection: ${perConn} msg/s`);

    results.push({
      tier: n,
      connected: actual,
      errors: n - actual,
      durationSecs: totalSecs,
      messagesSent: totalMessages,
      messagesReceived: 0,
      bytesSent: totalBytes,
      latency: null,
      extra: {
        msgs_per_sec: msgsPerSec,
        bytes_per_sec: bytesPerSec,
        per_conn_msgs: perConn,
      },
    });

    await protocol.closeAll(wsList);
    console.log();
  }

  // Print summary table
  if (results.length > 1) {
    console.log('  | Connections | Msg/s | MB/s | Per-conn msg/s | Errors |');
    console.log('  |-----------|-------|------|---------------|--------|');
    for (const r of results) {
      const rate = r.durationSecs > 0 ? r.messagesSent / r.durationSecs : 0;
      const mbps = r.durationSecs > 0 ? r.bytesSent / r.durationSecs / 1_000_000 : 0;
      const perConn = r.connected > 0 ? Math.round(rate / r.connected) : 0;
      const tierLabel = r.tier >= 1000 ? `${(r.tier / 1000).toFixed(0)}K` : `${r.tier}`;
      console.log(
        `  |  ${tierLabel.padEnd(8)} | ${fmtCount(Math.round(rate)).padEnd(8)} | ${mbps.toFixed(1).padEnd(6)} | ${String(perConn).padEnd(13)} | ${String(r.errors).padEnd(6)} |`
      );
    }
  }

  return results;
}

async function sendLoop(
  ws: protocol.WsConnection,
  msgBuf: Buffer,
  deadline: number,
): Promise<{ count: number; bytes: number; ws: protocol.WsConnection }> {
  let count = 0;
  let bytes = 0;
  const msgSize = msgBuf.length;

  return new Promise((resolve) => {
    function send() {
      while (Date.now() < deadline) {
        // Check backpressure — if buffered amount is too high, wait
        if ((ws as any).bufferedAmount > 16 * 1024 * 1024) {
          setTimeout(send, 1);
          return;
        }

        const ok = ws.send(msgBuf, { binary: false }, (err) => {
          if (err) {
            resolve({ count, bytes, ws });
          }
        });

        count++;
        bytes += msgSize;

        // Yield to event loop periodically
        if (count % 1000 === 0) {
          setImmediate(send);
          return;
        }
      }
      resolve({ count, bytes, ws });
    }

    send();
  });
}

File: benchmarks/rust-bench/src/tests/throughput.rs
use crate::config::Cli;
use crate::protocol::{self, WsStream};
use crate::report::{self, TierResult};
use crate::stats;
use futures_util::SinkExt;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use tokio_tungstenite::tungstenite::protocol::Message;

/// Test 3: Throughput Saturation
/// All connections blast messages as fast as possible. Finds peak msg/s.
pub async fn run(cli: &Cli) -> Vec<TierResult> {
    stats::print_header(
        "TEST 3: Throughput Saturation",
        &format!(
            "All connections send as fast as possible for {}s. Finds peak msg/s.",
            cli.duration
        ),
    );

    let tiers = cli.tiers_for(crate::config::TestName::Throughput);
    let token = crate::jwt::generate_bench_token(cli.secret.as_bytes(), "bench-user");
    let payload = protocol::build_payload(194); // Standard 194-byte trading payload
    let payload_len = payload.len() as u64;
    let mut results = Vec::new();

    for &n in &tiers {
        println!("\n  --- {} connections ---", n);

        let connections = protocol::connect_batch(
            &cli.host,
            cli.port,
            &token,
            n,
            cli.batch_size,
            "compression=false&protocol_version=1",
        )
        .await;

        if connections.is_empty() {
            println!("    FAILED: no connections");
            continue;
        }

        let actual = connections.len();
        let total_sent = Arc::new(AtomicU64::new(0));
        let total_bytes = Arc::new(AtomicU64::new(0));
        let duration = std::time::Duration::from_secs(cli.duration);

        // Start all senders simultaneously
        let start = tokio::time::Instant::now();
        let mut handles = Vec::with_capacity(actual);

        for ws in connections {
            let sent = total_sent.clone();
            let bytes = total_bytes.clone();
            let msg = payload.clone();
            let plen = payload_len;

            handles.push(tokio::spawn(async move {
                send_loop(ws, &msg, plen, duration, sent, bytes).await
            }));
        }

        let ws_list: Vec<WsStream> = futures_util::future::join_all(handles)
            .await
            .into_iter()
            .filter_map(|r| r.ok())
            .collect();

        let elapsed = start.elapsed().as_secs_f64();
        let sent = total_sent.load(Ordering::Relaxed);
        let bytes = total_bytes.load(Ordering::Relaxed);

        let msg_per_sec = sent as f64 / elapsed;
        let mb_per_sec = bytes as f64 / elapsed / 1_000_000.0;
        let per_conn = msg_per_sec / actual as f64;

        println!("    Duration: {:.2}s", elapsed);
        println!(
            "    Messages: {} ({}/s)",
            stats::fmt_num(sent),
            stats::fmt_rate(msg_per_sec)
        );
        println!(
            "    Bandwidth: {}/s",
            stats::fmt_bytes_per_sec(bytes as f64 / elapsed)
        );
        println!("    Per-connection: {:.0} msg/s", per_conn);

        results.push(TierResult {
            tier: n,
            connected: actual,
            errors: n - actual,
            duration_secs: elapsed,
            messages_sent: sent,
            messages_received: 0,
            bytes_sent: bytes,
            latency: None,
            extra: serde_json::json!({
                "msg_per_sec": msg_per_sec,
                "mb_per_sec": mb_per_sec,
                "per_conn_msg_s": per_conn,
            }),
        });

        protocol::close_all(ws_list).await;
        println!();
    }

    report::print_throughput_table(&results);
    results
}

async fn send_loop(
    mut ws: WsStream,
    payload: &str,
    payload_len: u64,
    duration: std::time::Duration,
    total_sent: Arc<AtomicU64>,
    total_bytes: Arc<AtomicU64>,
) -> WsStream {
    let deadline = tokio::time::Instant::now() + duration;
    let mut local_sent = 0u64;
    let mut local_bytes = 0u64;
    let msg = Message::Text(payload.to_string().into());

    while tokio::time::Instant::now() < deadline {
        if ws.send(msg.clone()).await.is_err() {
            break;
        }
        local_sent += 1;
        local_bytes += payload_len;

        // Flush periodically to avoid buffer bloat
        if local_sent.is_multiple_of(64) {
            let _ = ws.flush().await;
        }
    }

    total_sent.fetch_add(local_sent, Ordering::Relaxed);
    total_bytes.fetch_add(local_bytes, Ordering::Relaxed);
    ws
}
